### ISSUE TITLE
libbacktrace allocates ~131 mb the first time `libbacktrace::resolve()` is called and never frees it

### ISSUE THREAD

**Issue by Tomcc** [rust-lang/backtrace-rs#222]
I'm not sure if this is an issue in general, but it definitely was for me :)

tl;dr: I would like a way to tell `backtrace` to NOT keep its state around and allocate everything from scratch every time. I'm up for writing a PR myself!

long version:

I had an issue with a couple Rust services killing my low-mem (1gb) linux VMs by causing it to start swapping permanently, seemingly at random.
It turned out that the unassuming 30mb RAM processes could randomly grow permanently to 170mb or more, causing the VM to run out of memory.

I managed to repro the sequence of actions on Mac OS, so that let me trace the allocations with XCode Instruments, which is a pretty neat tool, and... it turns out the culprit is this crate!

<img width="1095" alt="Screen Shot 2019-07-07 at 8 49 00 AM" src="https://user-images.githubusercontent.com/553214/60769355-37021c80-a094-11e9-8b04-d48da68daf57.png">

Around the 1:40 mark, the program logged its first `failure::Error`, causing a backtrace resolve, causing 131 mb of persistent allocations. All the allocation in that image come from backtrace. For example
<img width="1677" alt="Screen Shot 2019-07-07 at 9 04 47 AM" src="https://user-images.githubusercontent.com/553214/60769513-62860680-a096-11e9-80a6-2fae209d2771.png">

The allocations never go away after that. Each new `resolve()` may allocate more tiny objects but never spikes as much as the first time.

My suspect is https://github.com/rust-lang/backtrace-rs/blob/master/src/symbolize/libbacktrace.rs#L260
The state is `static`, so it's probably what's keeping all those buffers forever alive. 
I assume that allocating a hundred mb or two forever may be worth the performance benefit on beefier machines, but in memory-constrained systems it's a problem.

My crappy 1gb VM would survive swapping until the trace is done (errors are very rare) but it can't survive keeping all that stuff around forever.
This is matters to bigger VMs too... for example, if you tried hosting dozens of containers on the same VM, they would slowly balloon in memory size as they encounter their first error, but it's super unlikely that they would all try to resolve an error at once.

I thought of just making it a #[cfg] option and submitting a PR, but I wanted to know if you think this makes sense to address and how you'd do it :)

If you want to see the original trace and have XCode, here it is:
[backtrace.trace.zip](https://github.com/rust-lang/backtrace-rs/files/3365694/backtrace.trace.zip)

**Comment by alexcrichton** [comment]
Yes currently this crate caches debuginfo in globals since it's so horribly slow to do it on all calls to creation of a backtrace. Having some sort of runtime option to clear that state seems reasonable but it would actually be somewhat difficult to implement on all platforms.

Can you try perhaps turning on the `gimli-symbolize` feature and see if the allocation amount goes down to a reasonable amount? Note though that libstd's `RUST_BACKTRACE=1` is also implemented through this crate, and that's much more difficulty to change than this crate.

**Comment by Tomcc** [comment]
> Yes currently this crate caches debuginfo in globals since it's so horribly slow to do it on all calls to creation of a backtrace. Having some sort of runtime option to clear that state seems reasonable but it would actually be somewhat difficult to implement on all platforms.
Yeah, that would also work for me! Fortunately I don't have a lot of paths that print out backtraces, so it should be doable.
If this is intended though, maybe it would be worth mentioning the memory cost in the docs? I would never have expected this behavior, not knowing the internals :)

> Can you try perhaps turning on the gimli-symbolize feature and see if the allocation amount goes down to a reasonable amount?
Sure, I'll try that.

> Note though that libstd's RUST_BACKTRACE=1 is also implemented through this crate, and that's much more difficulty to change than this crate.
At least in my case, that's fine because AFAIK `RUST_BACKTRACE` is only ever used when the program hit an unhandled panic and is about to be killed... so it doesn't matter that it won't free the debug info really.

**Comment by alexcrichton** [comment]
Adding docs seems reasonable to me! Most of this crate is just best effort on platforms anyway, so having a function which globally frees debuginfo and is implemented on a best-effort basis isn't entirely unreasonable as well.

**Comment by Tomcc** [comment]
@alexcrichton I tried `gimli-symbolize` and it's better, but not really... it allocates ~90mb, which is less, but it's inside a `thread_local`, so it can be more because unlike `libbacktrace`, it's going to be 90mb for each thread that resolves a backtrace.
My game logs errors to sentry from code that runs in a rayon threadpool, so as weird as it sounds, yes some weirdos do resolve backtraces on several threads :)

It also makes it harder to add that `free` function, given that you'd need to find all threadlocals, lock them, and free each one by one.

Also, I just learned about the `[patch]` feature in cargo... if the "no cache" behavior was a feature, would it be possible to write something like this in the root Cargo.toml:
```rs
[patch.crates-io.backtrace]
version = "0.3"
features = ["no-cache"]
```
and disable caching through all my dependencies? That seems less error prone than having to clean up after any dependency resolves a trace. In my program, both `failure` and `sentry` do that, for example.

**Comment by Tomcc** [comment]
I'm back with more bad news :)

I tried prototyping a change and... `libbacktrace` just doesn't support freeing the state at all. It's an opaque C object that can't be freed without understanding `libbacktrace` internals pretty well.
Not only that, I'm not sure that `libbacktrace` can free what it allocates at all... 
[I found this quote in `backtrace_free_unlocked`](https://github.com/ianlancetaylor/libbacktrace/blob/master/mmap.c#L72): 
```c
 /* Just leak small blocks.  We don't have to be perfect.  Don't put
     more than 16 entries on the free list, to avoid wasting time
     searching when allocating a block.  If we have more than 16
     entries, leak the smallest entry.  */
```
C, amirite? :)

Looks like gimli is the only option right now.

**Comment by alexcrichton** [comment]
Thanks for testing! I opened https://github.com/rust-lang/backtrace-rs/issues/225 since the gimli cache can be global rather than thread local (it's an old implementation!) and if you're interested in prototyping that and/or free support for gimli that seems reasonable to me!

**Comment by Tomcc** [comment]
Sure, I'll give it a shot... it might just take a looong time as this would be a side project of a side project!
First thing though, I tried [just recreating the `gimli` context every time](https://github.com/Tomcc/backtrace-rs/commit/2e2ecb75ca13c4862eb2f36f6f68ab356d36d6cc) and the crate still works, however the total memory as seen by Activity Monitor or `top` is still not going down... so before anything else I should make sure I catch everything that's allocating "forever".

**Comment by alexcrichton** [comment]
AFAIK that should catch everything, but it sounds like there's a bug somewhere although I'm not quite sure where

**Comment by Tomcc** [comment]
Should we close this issue now that the PR went in? I don't know... `libbacktrace` still has the same issue.

**Comment by alexcrichton** [comment]
Sounds good to me!
